examples/computer_vision/mnist_tf_layers/model_def.py

```diff
@@ -0,0 +1,131 @@
+"""
+An example showing how to use a graph defined in low-level TensorFlow APIs in Determined.
+
+We will be wrapping the TensorFlow graph in an Estimator and using Determined's EstimatorTrial.
+"""
+import tensorflow as tf
+import tensorflow_datasets as tfds
+
+from determined import estimator
+
+NUM_CLASSES = 10
+
+
+def calculate_logits(hparams, images: tf.Tensor, training: bool) -> tf.Tensor:
+    """This example assumes you already have something like this written for defining your graph."""
+    conv1 = tf.layers.conv2d(
+        inputs=tf.cast(images, tf.float32),
+        filters=hparams["n_filters_1"],
+        kernel_size=[5, 5],
+        padding="same",
+        activation=tf.nn.relu,
+    )
+    pool1 = tf.layers.max_pooling2d(inputs=conv1, pool_size=[2, 2], strides=2)
+
+    conv2 = tf.layers.conv2d(
+        inputs=pool1,
+        filters=hparams["n_filters_2"],
+        kernel_size=[5, 5],
+        padding="same",
+        activation=tf.nn.relu,
+    )
+    pool2 = tf.layers.max_pooling2d(inputs=conv2, pool_size=[2, 2], strides=2)
+    pool2_shape = pool2.get_shape().as_list()
+
+    pool2_flat = tf.reshape(pool2, [-1, pool2_shape[1] * pool2_shape[2] * pool2_shape[3]])
+    dense = tf.layers.dense(inputs=pool2_flat, units=512, activation=tf.nn.relu)
+
+    if training:
+        dropout = tf.layers.dropout(inputs=dense, rate=0.5)
+        logits = tf.layers.dense(inputs=dropout, units=NUM_CLASSES)
+    else:
+        logits = tf.layers.dense(inputs=dense, units=NUM_CLASSES)
+
+    return logits
+
+
+def calculate_loss(labels: tf.Tensor, logits: tf.Tensor) -> tf.Tensor:
+    """This example assumes you already have something like this written for defining your graph."""
+    return tf.reduce_mean(
+        tf.nn.sparse_softmax_cross_entropy_with_logits(labels=labels, logits=logits)
+    )
+
+
+def calculate_predictions(logits: tf.Tensor) -> tf.Tensor:
+    """This example assumes you already have something like this written for defining your graph."""
+    return tf.argmax(logits, axis=1)
+
+
+def calculate_error(predictions: tf.Tensor, labels: tf.Tensor) -> tf.Tensor:
+    """This example assumes you already have something like this written for defining your graph."""
+    correct = tf.cast(tf.equal(predictions, labels), tf.float32)
+    return 1 - tf.reduce_mean(correct)
+
+
+class MNistTrial(estimator.EstimatorTrial):
+    def __init__(self, context: estimator.EstimatorTrialContext) -> None:
+        self.context = context
+
+    def build_estimator(self) -> tf.estimator.Estimator:
+
+        # Define a model_fn which is the magic ingredient for wrapping a tensorflow graph in an
+        # Estimator.  The Estimator training loop will call this function with different modes to
+        # build graphs for either training or validation (or prediction, but that's not used by
+        # Determined).
+        #
+        # Read more at https://www.tensorflow.org/guide/estimator.
+
+        def model_fn(features, mode):
```

Comment:
something like this:

```
def make_model_fn(context):
    def model_fn(features, labels, mode):
        ....
     return model_fn
```

Reply:
I had it out but the `build_model_fn` was a bit silly how empty it was, and I wasn't crazy about the extra layer of function definitions.  Given that, do you still want it in a separate function?

No, the Estimator API actually introspects your function signature.  so you can ask for `labels` and it will be None or you can not ask for it and it's fine.

Reply:
I've seen some weird errors where Estimators will make full copy of the objects, not sure if it will apply here but I think it is generally good practice to have users keep it separate.

Reply:
ok, done.